tservice/src/seal.rs

```diff
@@ -71,6 +71,77 @@ impl<T: AsRef<[u8]>> Seal<Vec<u8>> for UnsealedData<T> {
     }
 }
 
+pub trait Unseal<T> {
+    /// The size in bytes needed to hold the unsealed MAC text
+    fn mac_text_len(&self) -> Result<usize>;
+
+    /// The length (in bytes) needed to hold the decrypted text
+    fn decrypted_text_len(&self) -> Result<usize>;
+
+    /// Unseal the data in `self`
+    fn unseal(&self) -> Result<UnsealedData<T>>;
+}
+
+impl<T: AsRef<[u8]>> Unseal<Vec<u8>> for SealedData<T> {
+    fn mac_text_len(&self) -> Result<usize> {
+        let result = unsafe {
+            mc_sgx_tservice_sys::sgx_get_add_mac_txt_len(
+                self.as_ref().as_ptr() as *const sgx_sealed_data_t
+            )
+        };
+        match result {
+            // Per the documentation, UINT32_MAX indicates an error
+            u32::MAX => Err(Error::Unexpected),
+            size => Ok(size as usize),
+        }
+    }
+
+    fn decrypted_text_len(&self) -> Result<usize> {
+        let result = unsafe {
+            mc_sgx_tservice_sys::sgx_get_encrypt_txt_len(
+                self.as_ref().as_ptr() as *const sgx_sealed_data_t
+            )
+        };
+        match result {
+            // Per the documentation, UINT32_MAX indicates an error
+            u32::MAX => Err(Error::Unexpected),
+            size => Ok(size as usize),
+        }
+    }
+
+    fn unseal(&self) -> Result<UnsealedData<Vec<u8>>> {
+        let data_length = self.decrypted_text_len()?;
+        let mac_length = self.mac_text_len()?;
+        let mut data = vec![0; data_length];
+        let mut mac = vec![0; mac_length];
+
+        let mac_pointer = match mac_length {
+            0 => ptr::null_mut(),
+            _ => mac.as_mut_ptr(),
+        };
+
+        let mut data_length = data_length as u32;
+        let mut mac_length = mac_length as u32;
```

Comment:
That these are being passed as `mut` to the FFI binding is giving me pause. I feel like we should return an internal error if `mac_length` or `data_length` has changed...

Reply:
I think this goes back to the common C api, of here is my buffer and it has room for X bytes, then the function fills the buffer with Y bytes replacing the size with the _used_ size.
```
     *      p_additional_MACtext_length - [IN/OUT] pointer to length of the plaintext data stream in bytes
     *                             If there is not additional data, this parameter should be zero.
```
This may be a failure on my part to [resize](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize) with the values afterward.

Or are you thinking that they shouldn't change since previously we got the needed length?

Reply:
https://github.com/mobilecoinfoundation/sgx/pull/127/commits/18bf2d20d0b8caf3d0dd6c81560ce40336c2faca added logic to error if the sizes differ from what SGX told us originally.